test-data/unit/fine-grained.test

```diff
@@ -8134,3 +8134,80 @@ x = 'no way'
 ==
 ==
 main:10: error: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+[case testLiteralFineGrainedVarConversion]
+import mod
+reveal_type(mod.x)
+[file mod.py]
+x = 1
+[file mod.py.2]
+from typing_extensions import Literal
+x: Literal[1] = 1
+[file mod.py.3]
+from typing_extensions import Literal
+x: Literal[1] = 2
+[out]
+main:2: error: Revealed type is 'builtins.int'
+==
+main:2: error: Revealed type is 'Literal[1]'
+==
+mod.py:2: error: Incompatible types in assignment (expression has type "Literal[2]", variable has type "Literal[1]")
+main:2: error: Revealed type is 'Literal[1]'
+
+[case testLiteralFineGrainedFunctionConversion]
+from mod import foo
+foo(3)
+[file mod.py]
+def foo(x: int) -> None: pass
+[file mod.py.2]
+from typing_extensions import Literal
+def foo(x: Literal[3]) -> None: pass
+[file mod.py.3]
+from typing_extensions import Literal
+def foo(x: Literal[4]) -> None: pass
+[out]
+==
+==
+main:2: error: Argument 1 to "foo" has incompatible type "Literal[3]"; expected "Literal[4]"
+
+[case testLiteralFineGrainedAlias]
+from mod import Alias
+a: Alias = 1
+[file mod.py]
+Alias = int
+[file mod.py.2]
+from typing_extensions import Literal
+Alias = Literal[1]
+[file mod.py.3]
+from typing_extensions import Literal
+Alias = Literal[2]
+[out]
+==
+==
+main:2: error: Incompatible types in assignment (expression has type "Literal[1]", variable has type "Literal[2]")
+
+[case testLiteralFineGrainedOverload]
+from mod import foo
+reveal_type(foo(4))
+[file mod.py]
+from typing import overload
+from typing_extensions import Literal
+@overload
+def foo(x: int) -> str: ...
+@overload
+def foo(x: Literal['bar']) -> int: ...
+def foo(x): pass
+[file mod.py.2]
+from typing import overload
+from typing_extensions import Literal
+@overload
+def foo(x: Literal[4]) -> Literal['foo']: ...
+@overload
+def foo(x: int) -> str: ...
+@overload
+def foo(x: Literal['bar']) -> int: ...
+def foo(x): pass
+[out]
+main:2: error: Revealed type is 'builtins.str'
+==
+main:2: error: Revealed type is 'Literal['foo']'
```

Comment:
These tests are indeed not very useful. What might be more interesting is to test the situations with "chained" assignments (where you have `y = x`, `z = y`, `w = z` etc, in different targets/modules) and where the type of the initial variable is changed, that causes a type error to appear with the last variable.

A similar idea is when you have a chain of function aliases where the last one provides a context for a function call. You can try playing with this.

Also maybe more interesting tests can be added when you will implement intelligent indexing and interaction with `Final`.